File: baekjoon/Java/P15652_N과_M(4).java
// 중복조합, 백트래킹

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Main {
	static List<List<Integer>> ans = new ArrayList<>();
	static int N, M;
	static StringBuilder sb = new StringBuilder();
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);

		N = sc.nextInt();
		M = sc.nextInt();

		dfs(0, 1, new ArrayList<>());

		System.out.print(sb.toString());
		sc.close();
	}

	public static void dfs(int n, int s, List<Integer> lst) {
		if (n == M) {
			// ans.add(new ArrayList<>(lst));
			for (int num : lst) {
				sb.append(num).append(" ");
			}
			sb.append("\n");
			return;
		}

		// s부터 N까지 반복
		for (int j = s; j <= N; j++) {
			lst.add(j);
			dfs(n + 1, j, lst);
			lst.remove(lst.size()-1);
		}
	}
}

File: baekjoon/Python/P15652_N과_M(4).py
"""
중복조합, 백트래킹
"""
def dfs(n, s, lst):
    if n==M:
        ans.append(lst)
        return
    
    for j in range(s, N + 1):
        dfs(n+1, j, lst + [j])


ans = []
N, M = map(int, input().split())
dfs(0, 1, [])
for lst in ans:
    print(*lst)
"""
from itertools import combinations_with_replacement

N, M = map(int, input().split())

li = [i for i in range(1, N + 1)]

ans = combinations_with_replacement(li, M)

for a in ans:
    print(*a)
"""
